Fix union field table handling

defunctzombie opened the pull request:
This fixes a bug with union field deserialization. While https://github.com/jkuszmaul/flatbuffers_reflection/pull/113 did fix the buggy handling of the typeIndex for union discriminator field reading, it did not address the same issue for `readTableLambda` and the use of `readVectorOfTables` in union handling logic.

In both cases, the logic to prepare a new `Table` instance(s) to correctly deserialize the union field  was using an incorrect "field type index". It was using the `fieldType.index()` value, which, for union fields, is actually the index into the enums list for the types that make up the union.

This change updates the union and union vector logic to use the correct specific types (based on discriminator value) when creating a table instance for further deserialization.

This change also surfaced that the union of struct handling code was broken. To make it clear that union of struct is not supported, the union of struct case is explicitly disallowed. A test is added showing it is explicitly disallowed.

---
@jkuszmaul here's another union parsing issue that I have fixed but don't quite understand where I went wrong the first time.

I've added a `this.readTableLambda2` showing the code that "works" to parse the message I have vs the original `this.readTableLamba` that does not.

One think I realized was that I was calling `this.readTableLambda(field, fieldType.index());` with `fieldType.index()` which is I think a similar bug to the discriminator value and I should have used the `typeIndex` from the caller. When I updated this line to `this.readTableLambda(field, typeIndex);` the logic in `readTableLambda` correctly identified `parentIsStruct` as _false_ (it was incorrectly true prior to that change), however, `elementIsStruct` was miss-identified as _true_ when it should actually be _false_.

I believe this last issue is because `this.getType(fieldType.index()).isStruct();`  is not correct when the `fieldType` refers to a union rather than an object (based on my reading of 
 https://github.com/google/flatbuffers/blob/master/reflection/reflection.fbs#L38).

Is it true that when reading a sub-table of a union field we _know_ the parent cannot be a struct and the element cannot be a struct? Or maybe because the element (the underlying union type _can_ be a struct) we still need this check but it needs to use the type index of the resolved union type rather than the `fieldType.index()`?

Is it even correct to construct the returned table as `new Table(table.bb, fieldType.index(), objectStart, elementIsStruct)` when `fieldType` refers to a Union type? I'm thinking no because `.index()` is not an index into objects but instead an index into enums?

defunctzombie commented:
@jkuszmaul Another strageness I've uncovered. I added an unused struct to the Unions.fbs file (https://github.com/jkuszmaul/flatbuffers_reflection/pull/115/commits/a3a568b6b0a33ec3e0fc71e1ec753e8cef7303ab#diff-d96cca0f36e7b272142e3cd750ced363bf653f7dc4914c5a1644da3622b9fb2aR4) as a "trap" for logic that might be using the wrong typeIndex.

However, this made the tests for union fail because they cannot find the discriminator field for the _vector_ fields. This is  happening because this code path is taken when reading the table trying to get the vector of discriminators: https://github.com/jkuszmaul/flatbuffers_reflection/blob/main/src/reflection.ts#L586

defunctzombie commented:
I was able to make more progress by building a "subtable reader" function per-enum type: https://github.com/jkuszmaul/flatbuffers_reflection/pull/115/commits/e24966bc4c09e63abc6f8b5a9193dff6d018a7ac#diff-2310f71825435beeb46aa425b9918c2fd06c4efbf50574f84761929c1557bee0R933 however I am still running into some sadness with "supports union of struct" test. I suspect my assumption about some field type or offset is still wrong somewhere.

defunctzombie commented on src/reflection.ts:
```diff
@@ -528,14 +530,11 @@ export class Parser {
-      fieldType.baseType() !== reflection.BaseType.Union
-    ) {
-      throw new Error("Field " + field.name() + " is not an object or union type.");
+    if (fieldType.baseType() !== reflection.BaseType.Obj) {
```
Removed _union_ type support from this function because `fieldType.index()` is not correct for the semantics here for union types

defunctzombie commented on src/reflection.ts:
```diff
@@ -867,17 +876,21 @@ export class Parser {
   // eslint-disable-next-line @foxglove/prefer-hash-private
-  private createUnionDeserializers(fieldType: reflection.Type, readDefaults: boolean) {
-    const unionDeserializers = new Map<number, (t: Table) => Record<string, any>>();
+  private createUnionDeserializers(
```
This now returns a tuple. The first value is a function to "read a sub-message" from the table. Previously this was the thing that was using `readTableLambda` and was not using the correct type index.

defunctzombie commented on src/reflection.ts:
```diff
@@ -813,7 +816,12 @@ export class Parser {
         }
-        result.push(deserializer(subTable));
+
+        const typeTable = deserializer[0](subTable);
```
I think this use of the deserializer response tuple might be wrong because the subTable is already the "message table" and does not need further processing to be read?

jkuszmaul commented:
I think I'm going to need to sit down and take a more careful look at this to actually understand what is going on, but just trying to address the specific questions:

> Is it true that when reading a sub-table of a union field we know the parent cannot be a struct and the element cannot be a struct?

A struct cannot have members that are unions (only scalars/enums, other structs, and fixed-length arrays). I would generally assume that you cannot have a union type that can be a struct, but I have not actually checked what flatbuffers actually supports.

> Or maybe because the element (the underlying union type can be a struct) we still need this check but it needs to use the type index of the resolved union type rather than the fieldType.index()?

Answering this properly will force me to actually figure out what is going on, but the short answer may just be that existing code was written more generically than it needed to be or was to support multiple use-cases.

> Is it even correct to construct the returned table as new Table(table.bb, fieldType.index(), objectStart, elementIsStruct) when fieldType refers to a Union type? I'm thinking no because .index() is not an index into objects but instead an index into enums?

Your intuition sounds correct; without the type of the actual object in question, we would have no way to actually parse the memory in question. If we wanted to be able to pass in the enum type index then we would need to have an interface that let you also pass in the specific object type index.

defunctzombie commented:
I've gotten this PR to a better place now. However I am still stuck on the _struct_ handling. I am doubting that the test is serializing correctly.

For the following schema
```
// This sets up a schema file with more structs than tables to ensure the union discriminator is
// correctly read from its parent table.

struct StructA {
  shine:double;
}

struct StructB {
  count:uint8;
}

union Kind { StructA, StructB }

table UnionStruct {
  kind:Kind;
}

root_type UnionStruct;
```

I get this serialization from the builder:

```
  Uint8Array(24) [
        12, 0, 0, 0, 8, 0, 11, 0,
        10, 0, 4, 0, 8, 0,  0, 0,
         7, 0, 0, 0, 0, 0,  2, 3
      ]
```

(Gotten from `console.log(builder.asUint8Array());`)

And this seems rather large to me for such a small schema...

defunctzombie commented:
@jkuszmaul After banging my head against the union of struct support I have instead opted to disable that functionality (by detecting if, within the union enums, any of them are a struct and throwing). I updated the "union of struct" test accordingly.

I added this test in https://github.com/jkuszmaul/flatbuffers_reflection/pull/113 and used "structs" as a way to create a schema where there were more structs than tables but I didn't actually _need_ union of struct support. What I should have done is the change I made in the Union schema for this PR by introducing a struct at the start of the schemas to flush out a bad index lookup. This previous PR fixed the typeIndex handling when reading the discriminator but failed to apply similar logic when reading the "field table" which is what this PR fixes.

What are your thoughts on the above and trying to make "union of struct" work? In theory it can work but my understanding of the flatbuffer wire format is not good enough to intuit through the table and vtable structure yet.

jkuszmaul commented:
I would much prefer not supporting union of structs to having broken behavior.

From reviewing the flatbuffers implementation, unions of structs are supported but the serialization format is poorly documented. For future reference, there is a very helpful annotation utility in flatbuffers to understand the contents and structure of a binary flatbuffer: https://github.com/google/flatbuffers/blob/master/tests/annotated_binary/README.md

Some discussion of how unions get serialized:

Given a schema
```
namespace aos.testing;

struct Abc {
  a:double;
}
struct Def {
  b:double;
  c:double;
}

table TestTable {
  d:int64 (id: 0);
}

union UnionOfStructs { Abc, Def, TestTable }

table TopLevelTable {
  union_field:UnionOfStructs (id: 1);
}
```

If we then construct the union once with the `Abc` struct with `a = 971` and then with the union set to a `TestTable` with `d = 118` then we get the below binary results.

```
[ RUN      ] Annotate.UnionOfTable
I20240702 22:25:26.245352    12 json_to_flatbuffer_test.cc:50] // Annotated Flatbuffer Binary
//
// Schema file: /home/james.kuszmaul/.cache/bazel/_bazel_james.kuszmaul/9107a858a519cd461ee7f5e02f7d6cbc/sandbox/linux-sandbox/176/execroot/shasta/_tmp/848f9e56e8d9a53789596e20a2e9404b/schema.bfbs
// Binary file: /home/james.kuszmaul/.cache/bazel/_bazel_james.kuszmaul/9107a858a519cd461ee7f5e02f7d6cbc/sandbox/linux-sandbox/176/execroot/shasta/_tmp/848f9e56e8d9a53789596e20a2e9404b/foo.bfbs

header:
  +0x00 | 0C 00 00 00             | UOffset32  | 0x0000000C (12) Loc: +0x0C | offset to root table `aos.testing.TopLevelTable`

vtable (aos.testing.TopLevelTable):
  +0x04 | 08 00                   | uint16_t   | 0x0008 (8)                 | size of this vtable
  +0x06 | 0A 00                   | uint16_t   | 0x000A (10)                | size of referring table
  +0x08 | 09 00                   | VOffset16  | 0x0009 (9)                 | offset to field `union_field_type` (id: 0)
  +0x0A | 04 00                   | VOffset16  | 0x0004 (4)                 | offset to field `union_field` (id: 1)

root_table (aos.testing.TopLevelTable):
  +0x0C | 08 00 00 00             | SOffset32  | 0x00000008 (8) Loc: +0x04  | offset to vtable
  +0x10 | 0C 00 00 00             | UOffset32  | 0x0000000C (12) Loc: +0x1C | offset to field `union_field` (union of type `TestTable`)
  +0x14 | 00                      | uint8_t[1] | .                          | padding
  +0x15 | 03                      | UType8     | 0x03 (3)                   | table field `union_field_type` (UType)

vtable (aos.testing.TestTable):
  +0x16 | 06 00                   | uint16_t   | 0x0006 (6)                 | size of this vtable
  +0x18 | 0C 00                   | uint16_t   | 0x000C (12)                | size of referring table
  +0x1A | 04 00                   | VOffset16  | 0x0004 (4)                 | offset to field `d` (id: 0)

table (aos.testing.TestTable):
  +0x1C | 06 00 00 00             | SOffset32  | 0x00000006 (6) Loc: +0x16  | offset to vtable
  +0x20 | 76 00 00 00 00 00 00 00 | int64_t    | 0x0000000000000076 (118)   | table field `d` (Long)
[       OK ] Annotate.UnionOfTable (0 ms)
[ RUN      ] Annotate.UnionOfStruct
I20240702 22:25:26.245555    12 json_to_flatbuffer_test.cc:61] // Annotated Flatbuffer Binary
//
// Schema file: /home/james.kuszmaul/.cache/bazel/_bazel_james.kuszmaul/9107a858a519cd461ee7f5e02f7d6cbc/sandbox/linux-sandbox/176/execroot/shasta/_tmp/848f9e56e8d9a53789596e20a2e9404b/schema.bfbs
// Binary file: /home/james.kuszmaul/.cache/bazel/_bazel_james.kuszmaul/9107a858a519cd461ee7f5e02f7d6cbc/sandbox/linux-sandbox/176/execroot/shasta/_tmp/848f9e56e8d9a53789596e20a2e9404b/foo.bfbs

header:
  +0x00 | 0C 00 00 00             | UOffset32  | 0x0000000C (12) Loc: +0x0C | offset to root table `aos.testing.TopLevelTable`

vtable (aos.testing.TopLevelTable):
  +0x04 | 08 00                   | uint16_t   | 0x0008 (8)                 | size of this vtable
  +0x06 | 0C 00                   | uint16_t   | 0x000C (12)                | size of referring table
  +0x08 | 0B 00                   | VOffset16  | 0x000B (11)                | offset to field `union_field_type` (id: 0)
  +0x0A | 04 00                   | VOffset16  | 0x0004 (4)                 | offset to field `union_field` (id: 1)

root_table (aos.testing.TopLevelTable):
  +0x0C | 08 00 00 00             | SOffset32  | 0x00000008 (8) Loc: +0x04  | offset to vtable
  +0x10 | 08 00 00 00             | UOffset32  | 0x00000008 (8) Loc: +0x18  | offset to field `union_field` (union of type `Abc`)
  +0x14 | 00 00 00                | uint8_t[3] | ...                        | padding
  +0x17 | 01                      | UType8     | 0x01 (1)                   | table field `union_field_type` (UType)

union (aos.testing.Abc.union_field):
  +0x18 | 00 00 00 00 00 58 8E 40 | double     | 0x408E580000000000 (971)   | struct field `aos.testing.Abc.a` (Double)
```

Regarding serialization, what this is showing us is that:
1. For unions of tables, the table instance looks identical to if there just weren't a union; i.e., in this case the `UnionOfStruct` output serializes identically to a schema that looks like:
  ```table TestTable {
  d:int64 (id: 0);
}
table TopLevelTable {
  table_field:TestField (id: 0);
  random_one_byte_field:uint8 (id: 1);
}```
2. For unions of structs, the way that unions get serialized is unique among flatbuffers concepts. In particular, the entry in the inline table data will be an offset (like it is for the table) rather than just being the struct data serialized inline (as it would be normally). That offset then points to the raw struct data, which is serialized the same way it would be normally (just inline). To the best of my knowledge, there is no other situation where you will ever have an offset to a struct (as distinct from vtable offsets).

jkuszmaul commented on src/reflection.ts:
```diff
@@ -842,10 +857,34 @@ export class Parser {
+      const deserializer = (table: Table) => {
+        const offsetToOffset = table.offset + table.bb.__offset(table.offset, fieldOffset);
+        if (offsetToOffset === table.offset) {
+          return undefined;
```
I seem to have preferred to use `null` to indicate an unpopulated value for other field types, can we keep that consistent?

defunctzombie replied:
Its not entirely clear. I would agree that there are some code paths that return null, but others seem to say they can return `| undefined`.

https://github.com/jkuszmaul/flatbuffers_reflection/blob/main/src/reflection.ts#L289
https://github.com/jkuszmaul/flatbuffers_reflection/blob/main/src/reflection.ts#L253
https://github.com/jkuszmaul/flatbuffers_reflection/blob/main/src/reflection.ts#L750

IMO, `undefined` is more idiomatic and natural. Some libraries like using `null` but I've always found it awkward. But it is your call so if you want `null` I can update the various types so it is clear what is expected.

jkuszmaul replied:
I'll be honest I have no idea why I made the type annotations on those indicate returning `undefined` rather than `null`....

In terms of stylistic choices I don't have a strong preference, although I think I have a slight preference towards `null` for this, following the logic in the first sentence of [undefined values have not been defined yet, whereas null values indicate intentional absence of a value.](https://dev.to/typescripttv/what-is-the-difference-between-null-and-undefined-5h76#:~:text=The%20convention%20in%20TypeScript%20is,intentional%20absence%20of%20a%20value.) (and I would argue in these cases the absence of a value is "intentional"). However I also see things that just generally discourage the use of `null` entirely and I don't mind adhering to common conventions.

We should definitely be consistent, though.

defunctzombie replied:
That link is kind of meh - it is true that you could use it semantically like that - but in practice no one cares. `undefined` means both and that is sufficient.

defunctzombie replied:
Want me to do that in this PR or follow-up?

jkuszmaul replied:
per separate conversation, any further changes will be separate.

defunctzombie commented on src/reflection.ts:
```diff
@@ -760,12 +754,24 @@ export class Parser {
-    const unionDeserializers = this.createUnionDeserializers(fieldType, readDefaults);
+    for (const [enumVal, type] of this.getUnionTypes(fieldType)) {
+      const specificTypeIndex = type.index();
+      if (specificTypeIndex < 0) {
```
This happens for the _none_ enum since it has no specific type. Instead of having these checks here, I've added a comment to this logic in `getUnionTypes` where we ignore the none value.

defunctzombie commented on src/reflection.ts:
```diff
@@ -802,18 +811,24 @@ export class Parser {
+
+        // In a vector of union, each "element" is an object until we re-visit struct
+        // support
+        nextTableOffset += typeSize(reflection.BaseType.Obj);
```
There might be some general form but I am not sure it would make it _much_ clearer because the current logic is doing the table iteration while going through the discriminator values (rather than reading all of the tables up front) so it kinda feels natural and clear that you are iterating the offset (note you also have to use the correct "specific type index" when building the subtable so it isn't only a matter of iterating with offsets.

jkuszmaul replied:
Fair enough.

defunctzombie commented on src/reflection.ts:
```diff
@@ -802,18 +811,24 @@ export class Parser {
+          table.bb,
+          specificTypeInfo.typeIndex,
+          table.bb.__indirect(nextTableOffset),
+          false /* elementIsStruct */,
```
I too suspect it would not be crazy - but I'm gonna vote we punt on this to a follow-up PR. I can take it on that but there's less urgency than this bugfix.

jkuszmaul replied:
Absolutely punt it to later (if ever).

defunctzombie commented on src/reflection.ts:
```diff
@@ -775,19 +781,22 @@ export class Parser {
-      if (tables == null) {
+      const offsetToOffset = table.offset + table.bb.__offset(table.offset, field.offset());
+      if (offsetToOffset === table.offset) {
         throw new Error(`Malformed message: missing vector table for field: ${field.name()}`);
```
🤔 I have added a new test where the vector of union types and discriminators is unset (at least I am not setting them explicitly - maybe the auto-gen code is still setting them to 0 length) and the test passed with no changes to this logic.

So the intent here is that if the offsetToOffset equals the table.offset then that means the field is not present? And you are saying we should detect that and return `undefined` iff discriminators.length === 0?

jkuszmaul replied:
Sigh, the not-quite-complete expressiveness of the object API always irritates me.

https://github.com/jameskuszmaul-brt/flatbuffers_reflection/commit/c140a693c5845b2051ddfb813b140a9abfa96295 modifies the test to actually construct the flatbuffer how I want

jkuszmaul reviewed (approved):
Some open questions about how we handle null/undefined, but deferring those to https://github.com/foxglove/flatbuffers_reflection/pull/20
